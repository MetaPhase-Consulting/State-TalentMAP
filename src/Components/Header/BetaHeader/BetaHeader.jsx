import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import FontAwesome from 'react-fontawesome';
import Skeleton, { SkeletonTheme } from 'react-loading-skeleton';
import PermissionsWrapper from '../../../Containers/PermissionsWrapper';
import EditContentButton from '../../EditContentButton';
import TextEditor from '../../TextEditor';
import { EMPTY_FUNCTION } from '../../../Constants/PropTypes';
import { homeBannerContentFetchData, homeBannerContentPatchData } from '../../../actions/homeBannerContent';
import { focusById } from '../../../utilities';

const EDIT_BUTTON_ID = 'edit-home-content';
const SUBMIT_BUTTON_ID = 'submit-home-content';

class BetaHeader extends Component {
  constructor(props) {
    super(props);
    this.patchData = this.patchData.bind(this);
    this.submit = this.submit.bind(this);
    this.cancel = this.cancel.bind(this);
    this.toggleEditor = this.toggleEditor.bind(this);
    this.state = {
      editorVisible: false,
    };
  }

  componentWillMount() {
    this.getData();
  }

  getData() {
    this.props.fetchData();
  }

  patchData(data) {
    this.props.patchData(data);
  }

  toggleEditor() {
    const { editorVisible } = this.state;
    this.setState({ editorVisible: !editorVisible }, () => {
      const elToFocus = this.state.editorVisible ? SUBMIT_BUTTON_ID : EDIT_BUTTON_ID;
      focusById(elToFocus, 0);
    });
  }

  submit(data) {
    this.props.patchData(data);
    this.setState({ editorVisible: false });
  }

  cancel() {
    this.setState({ editorVisible: false });
  }

  render() {
    const { data, isLoading } = this.props;
    const { editorVisible } = this.state;
    return (
      <div className="usa-banner tm-beta-header">
        <div className="usa-grid usa-banner-inner padded-main-content">
          {
            !editorVisible &&
              <div className="loader">
                <FontAwesome name="gears" />
                <SkeletonTheme color="#FAD980" highlightColor="#FDEFCC">
                  {data || <Skeleton width="50%" duration={1.8} />}
                </SkeletonTheme>
              </div>
          }
          {
            !isLoading && !editorVisible &&
              <PermissionsWrapper permissions="bidder">
                <EditContentButton onToggle={this.toggleEditor} id={EDIT_BUTTON_ID} />
              </PermissionsWrapper>
          }
          {
            editorVisible &&
              <TextEditor
                initialText={data}
                cancel={this.cancel}
                onSubmitText={this.submit}
                submitProps={{ id: SUBMIT_BUTTON_ID }}
              />
          }
        </div>
      </div>
    );
  }
}

BetaHeader.propTypes = {
  data: PropTypes.string,
  isLoading: PropTypes.bool,
  fetchData: PropTypes.func,
  patchData: PropTypes.func,
};

BetaHeader.defaultProps = {
  data: '',
  hasErrored: false,
  isLoading: false,
  fetchData: EMPTY_FUNCTION,
  patchData: EMPTY_FUNCTION,
};

const mapStateToProps = state => ({
  data: state.homeBannerContent,
  hasErrored: state.homeBannerContentHasErrored,
  isLoading: state.homeBannerContentIsLoading,
});

export const mapDispatchToProps = dispatch => ({
  fetchData: () => dispatch(homeBannerContentFetchData()),
  patchData: data => dispatch(homeBannerContentPatchData(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(BetaHeader);
