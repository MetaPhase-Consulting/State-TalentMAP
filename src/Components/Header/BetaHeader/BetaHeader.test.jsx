import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import { testDispatchFunctions } from '../../../testUtilities/testUtilities';
import BetaHeader, { mapDispatchToProps } from './BetaHeader';

describe('BetaHeader', () => {
  it('is defined', () => {
    const wrapper = shallow(<BetaHeader.WrappedComponent />);
    expect(wrapper).toBeDefined();
  });

  it('calls patchData on this.patchData()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<BetaHeader.WrappedComponent patchData={spy} />);
    wrapper.instance().patchData();
    sinon.assert.calledOnce(spy);
  });

  it('updates state on this.toggleEditor()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<BetaHeader.WrappedComponent patchData={spy} />);
    expect(wrapper.instance().state.editorVisible).toBe(false);
    wrapper.instance().toggleEditor();
    expect(wrapper.instance().state.editorVisible).toBe(true);
  });

  it('updates state and calls patchData on this.submit()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<BetaHeader.WrappedComponent patchData={spy} />);
    wrapper.instance().setState({ editorVisible: true });
    expect(wrapper.instance().state.editorVisible).toBe(true);
    wrapper.instance().submit();
    expect(wrapper.instance().state.editorVisible).toBe(false);
    sinon.assert.calledOnce(spy);
  });

  it('updates state on this.cancel()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<BetaHeader.WrappedComponent patchData={spy} />);
    wrapper.instance().setState({ editorVisible: true });
    expect(wrapper.instance().state.editorVisible).toBe(true);
    wrapper.instance().cancel();
    expect(wrapper.instance().state.editorVisible).toBe(false);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<BetaHeader.WrappedComponent />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  describe('mapDispatchToProps', () => {
    const config = {
      fetchData: [],
      patchData: [{}],
    };
    testDispatchFunctions(mapDispatchToProps, config);
  });
});
