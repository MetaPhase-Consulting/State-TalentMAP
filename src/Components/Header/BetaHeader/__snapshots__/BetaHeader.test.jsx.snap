// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetaHeader matches snapshot 1`] = `
<div
  className="usa-banner tm-beta-header"
>
  <div
    className="usa-grid usa-banner-inner padded-main-content"
  >
    <div
      className="loader"
    >
      <FontAwesome
        name="gears"
      />
      <t
        color="#FAD980"
        highlightColor="#FDEFCC"
      >
        <t
          circle={false}
          count={1}
          duration={1.8}
          height={null}
          width="50%"
          wrapper={null}
        />
      </t>
    </div>
    <Connect(PermissionsWrapper)
      permissions="bidder"
    >
      <EditContentButton
        id="edit-home-content"
        onToggle={[Function]}
      />
    </Connect(PermissionsWrapper)>
  </div>
</div>
`;
